File: TYHomework/ViewController/RegisterViewController/RSTextFieldKeyboardViewController.m
//
//  RSTextFieldKeyboardViewController.m
//  FITogether
//
//  Created by taoYe on 15/3/17.
//  Copyright (c) 2015年 closure. All rights reserved.
//

#import "RSTextFieldKeyboardViewController.h"

typedef void (^RSTextFieldKeyboardDoneAction)();

@interface RSTextFieldKeyboardViewController () <UITextFieldDelegate>

@property (nonatomic, weak) UITextField *target;
@property (nonatomic, strong) NSMutableArray *textFields;

@property (nonatomic, assign) BOOL floatingKeyboard;

@property (nonatomic, assign) RSTextFieldKeyboardDoneAction doneAction;
@property (nonatomic, assign) BOOL shouldFloatingUI;

@property (nonatomic, strong) UIMotionEffect *motion;

@property (nonatomic, assign) CGFloat fbixt;

@end

@implementation RSTextFieldKeyboardViewController

- (UIMotionEffect *)motion {
    if (!_motion) {
        CGFloat motionXMinValue = -20.0;
        CGFloat motionYMinValue = -20.0;
        CGFloat motionXMaxValue = 20.0;
        CGFloat motionYMaxValue = 20.0;
        UIInterpolatingMotionEffect *xAxis = [[UIInterpolatingMotionEffect alloc] initWithKeyPath:@"center.x" type:UIInterpolatingMotionEffectTypeTiltAlongHorizontalAxis];
        xAxis.minimumRelativeValue = @(motionXMinValue);
        xAxis.maximumRelativeValue = @(motionXMaxValue);
        
        UIInterpolatingMotionEffect *yAxis = [[UIInterpolatingMotionEffect alloc] initWithKeyPath:@"center.y" type:UIInterpolatingMotionEffectTypeTiltAlongHorizontalAxis];
        yAxis.minimumRelativeValue = @(motionYMinValue);
        yAxis.maximumRelativeValue = @(motionYMaxValue);
        UIMotionEffectGroup *group = [[UIMotionEffectGroup alloc] init];
        group.motionEffects = @[xAxis, yAxis];
        _motion = group;
    }
    return _motion;
}

- (void)setFbixt:(CGFloat)fbixt {
    if (fbixt < 1.0) {
        fbixt = 2.2;
    }
    _fbixt = fbixt;
}

- (void)textFieldDidBeginEditing:(UITextField *)textField {
    _target = textField;
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    _target = nil;
}

- (void)setTextFields:(NSMutableArray *)textFields {
    [self deapplyMotion];
    _textFields = textFields;
    for (UITextField *textField in _textFields) {
        textField.delegate = self;
    }
    [self applyMotion];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    int index = 0;
    for (int i = 0; i < _textFields.count; i++) {
        UITextField *tf = _textFields[i];
        if (textField == tf) {
            index = i;
            break;
        }
    }
    if (index == (_textFields.count - 1)) {
        [self tipGestureActive:nil];
        if (_doneAction != nil) {
            [self doneAction];
        }
    } else {
        [_textFields[index + 1] becomeFirstResponder];
    }
    return YES;
}

- (IBAction)tipGestureActive:(UITapGestureRecognizer *)sender {
    [_target resignFirstResponder];
    _target = nil;
}

- (void)keyboardWillShow:(NSNotification *)notification {
    if (self.floatingKeyboard) {
        return;
    }
    if (!self.shouldFloatingUI) {
        return;
    }
    self.floatingKeyboard = YES;
    NSDictionary *info = notification.userInfo;
   CGRect newFrame = [info[UIKeyboardFrameBeginUserInfoKey] CGRectValue];
    [UIView animateWithDuration:0.3 animations:^{
        self.view.frame = newFrame;
    }];
}

- (void)keyboardWillHide:(NSNotification *)notification {
    if (!self.floatingKeyboard) {
        return;
    }
    if (!self.shouldFloatingUI) {
        return;
    }
    self.floatingKeyboard = NO;
    [UIView animateWithDuration:0.3 animations:^{
        self.view.frame = self.view.bounds;
    }];
}

- (void)bindKeyboardNotification {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHide:) name:UIKeyboardWillHideNotification object:nil];
}

- (void)applyMotion {
    for (UIView *view in self.view.subviews) {
        [view addMotionEffect:self.motion];
    }
}

- (void)deapplyMotion {
    for (UIView *view in self.view.subviews) {
        [view removeMotionEffect:self.motion];
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self bindKeyboardNotification];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self deapplyMotion];
}


- (void)viewDidLoad {
    [super viewDidLoad];
    _fbixt = 2.2;
    [self.view addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tipGestureActive:)]];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: TYHomework/ViewController/RegisterViewController/RSTextFieldKeyboardViewController.swift
//
//  RyxTextFieldKeyboardViewController.swift
//  FITogether
//
//  Created by closure on 11/27/14.
//  Copyright (c) 2014 closure. All rights reserved.
//

import UIKit

public class RyxTextFieldKeyboardViewController : UIViewController, UITextFieldDelegate {
    private weak var _target: UITextField? = nil
    
    private var floatingKeyboard = false
    
    private var textFields:[UITextField]?
    
    public typealias RyxTextFieldKeyboardDoneAction = () -> Void
    
    public var doneAction: RyxTextFieldKeyboardDoneAction?
    
    public var shouldFloatingUI = true
    
    private lazy var motion: UIMotionEffect = {
        let motionXMinValue:CGFloat = -20.0
        let motionYMinValue:CGFloat = -20.0
        
        let motionXMaxValue:CGFloat = 20.0
        let motionYMaxValue:CGFloat = 20.0
        
        let xAxis = UIInterpolatingMotionEffect(keyPath: "center.x", type: UIInterpolatingMotionEffectType.TiltAlongHorizontalAxis)
        xAxis.minimumRelativeValue = motionXMinValue
        xAxis.maximumRelativeValue = motionXMaxValue
        
        let yAxis = UIInterpolatingMotionEffect(keyPath: "center.y", type: UIInterpolatingMotionEffectType.TiltAlongHorizontalAxis)
        yAxis.minimumRelativeValue = motionYMinValue
        yAxis.maximumRelativeValue = motionYMaxValue
        
        let group = UIMotionEffectGroup()
        group.motionEffects = [xAxis, yAxis]
        return group
    }()
    
    public func textFieldDidBeginEditing(textField: UITextField) {
        _target = textField
    }
    
    public func textFieldDidEndEditing(textField: UITextField) {
        _target = nil
    }
    
    public func setTextFields(textFields: [UITextField]) {
        deapplyMotion()
        self.textFields = textFields
        for textField in self.textFields! {
            textField.delegate = self
        }
        applyMotion()
    }
    
    public func textFieldShouldReturn(textField: UITextField) -> Bool {
        if let index = find(self.textFields!, textField) {
            if index + 1 == self.textFields!.count {
                tipGestureActive(nil)
                if self.doneAction != nil {
                    self.doneAction!()
                }
            } else {
                self.textFields![index + 1].becomeFirstResponder()
            }
        }
        return true
    }
    
    @IBAction func tipGestureActive(sender: UITapGestureRecognizer?) {
        _target?.resignFirstResponder()
        _target = nil
    }
    
    public var fbixt: CGFloat = 2.2 {
        didSet {
            if fbixt < 1.0 {
                fbixt = 2.2
            }
        }
    }
    
    func keyboardWillShow(notification: NSNotification) {
        if self.floatingKeyboard {
            return
        }
        
        if !self.shouldFloatingUI {
            return
        }
        
        self.floatingKeyboard = true
        let info = notification.userInfo
        let kbSize = info![UIKeyboardFrameBeginUserInfoKey]!.CGRectValue().size
        var newFrame = self.view.frame
        newFrame.origin.y -= (kbSize.height / fbixt)
        UIView.animateWithDuration(0.3, animations: { () -> Void in
            self.view.frame = newFrame
        })
    }
    
    func keyboardWillHide(notification: NSNotification) {
        if !self.floatingKeyboard {
            return
        }
        if !self.shouldFloatingUI {
            return
        }
        self.floatingKeyboard = false
        UIView.animateWithDuration(0.3, animations: { () -> Void in
            self.view.frame = self.view.bounds
        })
    }
    
    func bindKeyboardNotification() {
        let center = NSNotificationCenter.defaultCenter()
        center.addObserver(self, selector: Selector("keyboardWillShow:"), name: UIKeyboardWillShowNotification, object: nil)
        center.addObserver(self, selector: Selector("keyboardWillHide:"), name: UIKeyboardWillHideNotification, object: nil)
    }
    
    func unbindKeyboardNotification() {
        let center = NSNotificationCenter.defaultCenter()
        center.removeObserver(self)
    }
    
    func applyMotion() {
        for view in self.view.subviews {
            view.addMotionEffect(self.motion)
        }
    }
    
    func deapplyMotion() {
        for view in self.view.subviews {
            view.removeMotionEffect(self.motion)
        }
    }
    
    public override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        bindKeyboardNotification()
        TalkingData.beginTrack(self.dynamicType)
    }
    
    public override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        deapplyMotion()
        unbindKeyboardNotification()
        TalkingData.endTrack(self.dynamicType)
    }
    
    public override func viewDidLoad() {
        super.viewDidLoad()
        let gesture = UITapGestureRecognizer(target: self, action: Selector("tipGestureActive:"))
        self.view.addGestureRecognizer(gesture)
    }
}

